### DISCUSSION TITLE
Installation experience

### DISCUSSION THREAD

**Discussion by minexew** [Nefaro/gnoll#2] [Ideas]
Hi, I was checking out your thread on Steam forums and I realized how bad the installation experience must be for non-developers.
I could look into producing a more adequate end-user packaging. No NASA-tier technology, but something to get the job done. Would you be interested?

**Comment by Nefaro** [comment]
Sure. Whatever makes the installation easier
And yes, it's bit complicated, even for myself :/

**Reply by minexew** [reply]
Ok, my idea (yet to be prototyped) is this:
- the development flow will not be changed
- to create a release, we run the usual build flow, and at the end generate a binary diff between the retail game and the patched EXE
- we make a simple 'installer' tool for the end-users to apply this binary diff, either replacing the game executable in-place (for Steam Workshop) or creating a separate one (GnoMod.exe or whatever)

**Reply by minexew** [reply]
For the installer, I think it would make most sense to also use C#, but I'm not sure about the GUI framework, since we probably want to keep the door open for a macOS/Linux port.

**Reply by Nefaro** [reply]
A binary diff-patch for the exe... That's interesting idea, I personally have no experience with that.
I mean, sure, a GUI for installer is much easier than anything else we have currently. This would also allow just to ship the compiled .dll for the mods... maybe even have a external enable-disable functionality etc. (which should also ... probably... exist ingame).

And yes, I like the idea of mac/*nix port.

However, I have no idea about C# ecosystem, so I don't have any suggestions. This here is my first time working on a C# project. I also have no idea, if there are any implications on having a pre-compiled .dll files and them breaking in different Windows version (say ... 7 or 11) or so.

I mean, if you can do it, I will not stand in your way and the users would certainly be thankful. 
Edit: Distributing a pre-compiled solution is a very exciting thought, just to sum up

**Reply by minexew** [reply]
I've done some UI mock-ups
![image](https://user-images.githubusercontent.com/1880105/140217255-716c7ba6-a656-4434-9668-dcd8b1eb9623.png)

**Reply by minexew** [reply]
Slightly updated

![image](https://user-images.githubusercontent.com/1880105/141957034-16a677cc-2760-4b83-a001-6fcc03898f0d.png)

**Reply by minexew** [reply]
@Nefaro do you have any graphics (preferably non-copyrighted) that I could use as the logo/banner?

**Reply by Nefaro** [reply]
Oh, that's an interesting question... I haven't even thought about any graphics. Currently, no, need to think about it.
For the first iteration, maybe even just the word "Gnoll" (as an image) would be good enough ?

**Reply by minexew** [reply]
Yeah, I can use a placeholder as well. Perhaps someone from the community will step up later :)

If you could make a patch for Gnoll 1.6 for me, I would try to produce a minimal first release. Unfortunately I don't have the python/.NET machinery configured at the moment, and more importantly, I don't own Gnomoria on Steam, so I can't make the patch myself.

The patch should be generated using [xdelta](https://github.com/jmacd/xdelta-gpl/releases/tag/v3.1.0) like this: `xdelta3 -S -s Gnomoria.exe GnoMod.exe Gnoll1.6.xdelta`

**Reply by Nefaro** [reply]
Sure. Pushed up a commit, put the diffs (1.6 and 1.7) into "./patch" directory.

**Reply by minexew** [reply]
Cool! If you're not afraid to run a random binary from the internet, you can try this out: https://github.com/minexew/gnoll/releases/tag/G1.7pre1

**Reply by Nefaro** [reply]
On the first try I didn't read, what was written so I pasted my path 2 times.
On the second try though, I'm not quite sure which path it's complaining about

![image](https://user-images.githubusercontent.com/1470511/142077664-4e5036aa-217d-4661-9448-19d57383ba45.png)

**Reply by Nefaro** [reply]
However, it did write the modified .exe and as far as I can tell, it works. 
This is so awesome. This will make the installation process so so much better.

**Reply by minexew** [reply]
Thanks for the feedback!
I'm not sure what went wrong in the second case, but I know where to look. It's related to writing a file called gnoll-version.json, so that the installer remembers what it has done.

**Reply by Nefaro** [reply]
It did generate a "gnoll-version.json.tmp" file. I assume, it had trouble renaming it or something.

**Reply by minexew** [reply]
Yup. This file is written first and then atomically moved to 'gnoll-version.json'. But it seems that `File.Replace` is not the correct function for the latter.

**Reply by minexew** [reply]
I have tried to fix the previous error and add a GUI: https://github.com/minexew/gnoll/releases/tag/G1.7pre2

**Reply by minexew** [reply]
WinForms turns out to be quite nice to work with, but the UI still needs a lot of polish.

**Reply by Nefaro** [reply]
It's a start. That's more UI than the previous build cycle. 
This time the patching worked without a trouble, version detection as well. Nothing crashed.
Very exciting stuff

**Reply by Nefaro** [reply]
Thank you for this installer GUI! I will play around with it during weekend... I already managed to crash it ... and it doesn't like my scaled display. 
But a single click install is sooo much better than all the fiddling with python and the framework.
I really like it

**Reply by minexew** [reply]
Yeeah, I'll have to figure out how to deal with Hi-DPI. I don't use a 4K screen myself, so I haven't noticed if there are any problems.

Have you been able to build the installer with your own payloads (patches)? I wrote some brief instructions [here](https://github.com/Nefaro/gnoll/tree/master/Installer#payloads), but in an ideal world this process should be integrated into the Python task executor, rather than having to do it manually.

**Reply by Nefaro** [reply]
No, not yet. I only managed to surf around the code a bit, to see how it works. Will have more time on weekend (maybe) and will try out the payload thing as well.

**Reply by Nefaro** [reply]
Tried out, all works :)  Very exciting
![image](https://user-images.githubusercontent.com/1470511/146656899-d2ded00f-a7bb-43b9-928d-093fa20600f1.png)

A few things I'd like to still do:
* Load patches from a folder (this means the patches can be published/released without updating the installer)
* Include a UI element to also install mods (this allows for the possibility to distribute a package or a .zip with mods and ask the installer to put them in the right place
* Then I need to figure out how to make a release in Github... and how to package it together (via Github)
* And few minor things (like how to incorporate the installer build number into the app itself)

In future I see 3 types of releases:
* Full (installer + latest patch + mods)
* Patch (latest patch + mods)
* Mod (just all mods)

I will work on all of this in the coming weeks, I'll be having a short vacation and will probably have more time.

**Reply by minexew** [reply]
I just wanted to note that the original goal was to distribute the installer as a single executable -- it should be possible with .NET to somehow bundle the dependent libraries. That's why I also embedded the patches as resources.

However, when also shipping a bunch of mods, this becomes really inconvenient to build, so it makes sense to just distribute a bunch of files in a ZIP.

Eventually, it should be possible to use some "external" bundler that generates a single executable, transparently extracting everything into a temporary directory before executing InstallerGUI.exe, but it's just nice-to-have, not essential.

**Reply by Nefaro** [reply]
Yeah, I originally also thought to distribute everything as a single pack... but now I'm thinking, that I personally don't see much need to upgrade the installer that often. Once it works, it should be more or less stable and ideally the mods are those that change, as such distributing them at different times makes more sense.

Then again, this all relies upon someone using it and building new mods etc. That might not even happen ...

**Reply by Nefaro** [reply]
After having played with the installer a bit more, the mod management is a bit ambitious idea, specially for a first release.
I did add loading patches from a folder (via metadata info file), also added displaying standalone version (at least latest) in the UI. 
For me this is already good enough. Also, forgot to mention, I like the banner, it's nice.

In the following days will figure out how to do a .zip based release for the installer and then for the mods and then, this should tear down the huge installation barriers that are present at the moment.

**Reply by Nefaro** [reply]
I made a release for the first pack: https://github.com/Nefaro/gnoll/releases/tag/full_v1.0.1
And had to fix it already, since I forgot GoG ...

**Comment by Nefaro** [comment]
Worked a bit on the UI to mitigate the scaling issue. 
I have a 2k (1440p) display and got it to work consistently ... more or less. The bottom line doesn't want to play ball. 
However made it so, that up to 150% should be fine. Bigger scaling ... has still problems.
Examples of 100% and 150%
![image](https://user-images.githubusercontent.com/1470511/146649556-b1c6cb5b-4fd9-4eae-bc60-ec1d26ff04d2.png)
![image](https://user-images.githubusercontent.com/1470511/146649745-63a84e7d-a95f-4386-ae73-220b07e65d4b.png)

The original issue was that the UI was jumping around: it started as scaled but when I clicked on the "Browse ..." button, it re-scaled itself down from 150% to 100%. Spent too much time on this, though I don't think we are much better of than before...

Added the toolbar ico though...

**Reply by Nefaro** [reply]
welp... after testing it some more... it's not an issue... after proper compile and running from outside of Visual Studio, the 100% one is taken and scaled up... Visual Studio is basically lying to me ...
